rhizome/lib/storage_key_encryption.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require_relative "common"
+require "openssl"
+require "base64"
+
+class StorageKeyEncryption
+  def initialize(key_encryption_cipher)
+    @key_encryption_cipher = key_encryption_cipher
+  end
+
+  def write_encrypted_dek(key_file, data_encryption_key)
+    File.write(key_file, JSON.pretty_generate({
+      cipher: data_encryption_key[:cipher],
+      key: wrap_key(data_encryption_key[:key]),
+      key2: wrap_key(data_encryption_key[:key2])
+    }))
+  end
+
+  def read_encrypted_dek(key_file)
+    data_encryption_key = JSON.parse(File.read(key_file))
+    {
+      cipher: data_encryption_key["cipher"],
+      key: unwrap_key(data_encryption_key["key"]),
+      key2: unwrap_key(data_encryption_key["key2"])
+    }
+  end
+
+  def wrap_key(key)
+    cipher = OpenSSL::Cipher.new(@key_encryption_cipher["algorithm"])
+    cipher.encrypt
+    cipher.key = Base64.decode64(@key_encryption_cipher["key"])
+    cipher.iv = Base64.decode64(@key_encryption_cipher["init_vector"])
+    cipher.auth_data = @key_encryption_cipher["auth_data"]
+    [
+      Base64.encode64(cipher.update(key) + cipher.final),
+      Base64.encode64(cipher.auth_tag)
+    ]
+  end
+
+  def unwrap_key(encrypted_key)
+    decipher = OpenSSL::Cipher.new(@key_encryption_cipher["algorithm"])
+    decipher.decrypt
+    decipher.key = Base64.decode64(@key_encryption_cipher["key"])
+    decipher.iv = Base64.decode64(@key_encryption_cipher["init_vector"])
+    decipher.auth_data = @key_encryption_cipher["auth_data"]
+    decipher.auth_tag = Base64.decode64(encrypted_key[1])
```

Comment:
in general, when working with auth_tag, should we make pains to be more robust to this? https://github.com/ruby/openssl/issues/63